File: 823-binary-trees-with-factors/823-binary-trees-with-factors.py
class Solution:
    def numFactoredBinaryTrees(self, arr: List[int]) -> int:
        dp = Counter(arr)
        self.ans = 0 
        arr.sort()
        
        def recurse(index):
            if index >= len(arr):
                return 0
            parent = arr[index]
            for j in range(index):
                child1 = arr[j]
                child2 = parent /child1
                if child2 in dp:
                    dp[parent] += dp[child1] * dp[child2]
                    
            recurse(index+1)
            self.ans += dp[parent]
            return dp[parent]
        
        recurse(0)
        return self.ans% (10**9+7)

File: 823-binary-trees-with-factors/823-binary-trees-with-factors.cpp
class Solution {
    long int ans = 0;
    unordered_map<long int, long int> dp;
    int mod = pow(10, 9) + 7;
public:
    int numFactoredBinaryTrees(vector<int>& arr) {
        
        for (int i=0; i < arr.size(); i++){
            dp[arr[i]] = 1;
        }
        
        sort(arr.begin(), arr.end());
        recurse(0, arr);
        return ans % mod;
        
    }
    
    long int recurse(int index, vector<int>& arr){
            if (index >= arr.size()){
                return 0;
            }
            
            int parent = arr[index];
            
            for (int j=0; j < index; j++){
                int child1 = arr[j], child2 = parent / double(child1);
                if (parent % child1 != 0) continue;
                
                dp[parent] += (dp[child1] * dp[child2]);
                
            }
            recurse(index+1, arr);
            
            ans += dp[parent];
            return dp[parent];
        }
};